stan/math/prim/mat/prob/poisson_log_glm_lpmf.hpp

```diff
@@ -77,62 +85,59 @@ typename return_type<T_x, T_alpha, T_beta>::type poisson_log_glm_lpmf(
   if (!include_summand<propto, T_x, T_alpha, T_beta>::value)
     return 0.0;
 
-  Matrix<T_partials_return, Dynamic, 1> beta_dbl(M, 1);
-  {
-    scalar_seq_view<T_beta> beta_vec(beta);
-    for (size_t m = 0; m < M; ++m) {
-      beta_dbl[m] = value_of(beta_vec[m]);
-    }
-  }
+  const auto& x_val = value_of_rec(x);
+  const auto& y_val = value_of_rec(y);
+  const auto& beta_val = value_of_rec(beta);
+  const auto& alpha_val = value_of_rec(alpha);
 
-  Matrix<T_partials_return, Dynamic, 1> theta_dbl = value_of(x) * beta_dbl;
+  const auto& y_val_vec = as_column_vector_or_scalar(y_val);
+  const auto& beta_val_vec = as_column_vector_or_scalar(beta_val);
+  const auto& alpha_val_vec = as_column_vector_or_scalar(alpha_val);
 
-  Matrix<T_partials_return, Dynamic, 1> theta_derivative(N);
-  T_partials_return theta_derivative_sum = 0;
-  scalar_seq_view<T_alpha> alpha_vec(alpha);
-  scalar_seq_view<T_y> y_seq_view(y);
-  for (size_t n = 0; n < N; ++n) {
-    // Compute the derivative wrt theta.
-    double theta_dbl_n = theta_dbl[n] + value_of(alpha_vec[n]);
-    double exp_theta_n = exp(theta_dbl_n);
-    theta_derivative[n] = y_seq_view[n] - exp_theta_n;
-    if (!is_vector<T_alpha>::value)
-      theta_derivative_sum += theta_derivative[n];
+  Matrix<T_partials_return, Dynamic, 1> theta = x_val * beta_val_vec;
+  theta.array() += as_array_or_scalar(alpha_val_vec);
 
-    // Compute the log-density.
-    if (!(theta_dbl_n == -std::numeric_limits<double>::infinity()
-          && y_seq_view[n] == 0)) {
-      if (include_summand<propto>::value)
-        logp -= lgamma(y_seq_view[n] + 1.0);
-      if (include_summand<propto, T_partials_return>::value)
-        logp += y_seq_view[n] * theta_dbl_n - exp_theta_n;
+  Matrix<T_partials_return, Dynamic, 1> theta_derivative
+      = as_array_or_scalar(y_val_vec) - exp(theta.array());
+  double theta_derivative_sum = theta_derivative.sum();
+  if (!std::isfinite(theta_derivative_sum)) {
+    check_finite(function, "Weight vector", beta);
+    check_finite(function, "Intercept", alpha);
+    check_finite(function, "Matrix of independent variables", theta);
+  }
+  if (include_summand<propto>::value) {
+    if (is_vector<T_y>::value) {
+      logp -= sum(lgamma(as_array_or_scalar(y_val_vec) + 1.0));
```

Comment:
[optional]
Don't need the `.0` after `1`---the compiler will sort out the literals.